test/onnx/verify/reducesum_variable_dynamic_axes_test.cpp

```diff
@@ -0,0 +1,48 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <migraphx/register_target.hpp>
+#include <migraphx/verify.hpp>
+#include <onnx_test.hpp>
+
+TEST_CASE(reducesum_variable_dynamic_axes_test)
+{
+    using namespace migraphx;
+    onnx_options options;
+    const std::vector<shape::dynamic_dimension> axes_dims{{0, 3}};
+    options.map_dyn_input_dims["axes"] = axes_dims;
+    program p = parse_onnx("reducesum_variable_dynamic_axes_verify_test.onnx", options);
+    p.compile(make_target("ref"));
+
+    parameter_map pm;
+    shape x_shape{shape::float_type, {2, 2, 2}};
+    std::vector<float> x(x_shape.elements());
+    std::iota(x.begin(), x.end(), 0);
+    pm["x"] = argument(x_shape, x.data());
+
+    std::vector<int64_t> axes{1};
+    pm["axes"] = argument(shape{shape::int64_type, {1}}, axes.data());
+
+    auto result = p.eval(pm);
```

Comment:
@pfultz2 @CharlieL7 

This test demonstrates the Where issue that crops up([here is the code path it tests](https://github.com/ROCm/AMDMIGraphX/pull/2120/files#diff-e2bd4cc9d4b1d5365a82126bf3a7bd6a4b317c30c4a2251cf801b9c3ae72fcd3R87)) we discussed last week.

All three input shapes of Where being the same({1, 2}, {1, 2}, {1, 2}) parsing will succeed.
However, when attempting to evaluate the graph issues pop up.
The shapes being reified, the all axes reduction will have an output shape of {1, 1, 1}, the other one {2, 1, 2}, causing [this check to fail](https://github.com/migraphx-benchmark/AMDMIGraphX/blob/develop/src/include/migraphx/op/where.hpp#L45).

Reply:
The `compute` method to `where` needs to be updated to remove the `dyn_output` usage so the compute shape can happen lazily. For scalar condition we can pick either branch regardless of the shape difference:

```cpp
    argument compute(shape output_shape, std::vector<argument> args) const
    {
        if(args[0].scalar())
            return args[args[0].at<bool>() ? 1 : 2].copy();
        if(output_shape.dynamic())
            output_shape = compute_shape(to_shapes(args));
        argument result{output_shape};
        visit_all(result, args[1], args[2])([&](auto output, const auto x, const auto y) {
            args[0].visit([&](const auto condition) {
                par_for(dyn_out.computed_shape.elements(),
                        [&](auto i) { output[i] = condition[i] ? x[i] : y[i]; });
            });
        });
        return result;
    }
```